src/metrics.rs

```diff
@@ -0,0 +1,306 @@
+    fn new(writer: Arc<MetricsWriter + Send + Sync>, write_frequency: Duration) -> Self {
+        let (sender, receiver) = channel::<MetricsCommand>();
+        thread::spawn(move || Self::run(receiver, writer, write_frequency));
+        MetricsAgent { sender }
+    }
+
+    fn run(
+        receiver: Receiver<MetricsCommand>,
+        writer: Arc<MetricsWriter>,
+        write_frequency: Duration,
+    ) {
+        trace!("run: enter");
+        let mut last_write_time = Instant::now();
+        let mut points = Vec::new();
+
+        loop {
+            match receiver.recv_timeout(write_frequency / 2) {
+                Ok(cmd) => match cmd {
+                    MetricsCommand::Flush(barrier) => {
+                        debug!("metrics_thread: flush");
+                        if !points.is_empty() {
+                            writer.write(points);
+                            points = Vec::new();
+                            last_write_time = Instant::now();
+                        }
+                        barrier.wait();
+                    }
+                    MetricsCommand::Submit(point) => {
+                        debug!("run: submit {:?}", point);
+                        points.push(point);
+                    }
+                },
+                Err(RecvTimeoutError::Timeout) => {
+                    trace!("run: receive timeout");
+                }
+                Err(RecvTimeoutError::Disconnected) => {
+                    debug!("run: sender disconnected");
+                    break;
+                }
+            }
+
+            let now = Instant::now();
+            if now.duration_since(last_write_time) >= write_frequency {
+                if !points.is_empty() {
+                    debug!("run: writing {} points", points.len());
+                    writer.write(points);
+                    points = Vec::new();
+                    last_write_time = now;
+                }
+            }
+        }
+        trace!("run: exit");
+    }
+
+    pub fn submit(&self, mut point: influxdb::Point) {
+        if point.timestamp.is_none() {
+            point.timestamp = Some(timing::timestamp() as i64);
+        }
+        debug!("Submitting point: {:?}", point);
+        self.sender.send(MetricsCommand::Submit(point)).unwrap();
+    }
+
+    pub fn flush(&self) {
+        debug!("Flush");
+        let barrier = Arc::new(Barrier::new(2));
+        self.sender
+            .send(MetricsCommand::Flush(Arc::clone(&barrier)))
+            .unwrap();
+
+        barrier.wait();
+    }
+}
+
+impl Drop for MetricsAgent {
+    fn drop(&mut self) {
+        self.flush();
+    }
+}
+
+fn get_singleton_agent() -> Arc<Mutex<MetricsAgent>> {
+    static INIT: Once = ONCE_INIT;
+    static mut AGENT: Option<Arc<Mutex<MetricsAgent>>> = None;
+    unsafe {
+        INIT.call_once(|| AGENT = Some(Arc::new(Mutex::new(MetricsAgent::default()))));
+        match AGENT {
+            Some(ref agent) => Arc::clone(&agent),
```

Comment:
The rest of the codebase writes that as `agent.clone()`. Can you follow suit? And if you find `Arc::clone()` easier to read add an issue to change them all? I'm not opposed.

Also, are you sure that `ref` was needed?  rustc 1.27 removed the need for most of those.

Reply:
Some online doc told me that `Arc::clone()` is more idiomatic, advice that as I noob I gobbled up as gospel.  `.clone()` wfm too, I have no opinion yet.

Yeah the `ref` is needed, without I get a
```
    |
164 |         match AGENT {
    |               ^^^^^ cannot move out of static item
165 |             Some(agent) => agent.clone(),
    |                  ----- hint: to prevent move, use `ref agent` or `ref mut agent`
```